lib/storage/index.js

```diff
@@ -0,0 +1,266 @@
+import _ from 'underscore';
+import PlatformStorage from './platforms';
+import MemoryOnlyProvider from './providers/MemoryOnlyProvider';
+import InstanceSync from './InstanceSync';
+import * as Logger from '../Logger';
+
+let storageProviders = [PlatformStorage, MemoryOnlyProvider];
+let finishInitalization;
+const initPromise = new Promise((resolve) => {
+    finishInitalization = resolve;
+});
+let shouldKeepInstancesSync = false;
+
+/**
+ * Runs an operation consecutively on all storage providers
+ * @param {String} operationName
+ * @param {Array<any>} params - the parameters to pass to the operation
+ * @return {Promise<void>}
+ */
+function runConsecutively(operationName, ...params) {
+    const runOperation = (providerIndex) => {
+        const provider = storageProviders[providerIndex];
+
+        return new Promise((resolve, reject) => {
+            provider[operationName](...params)
+                .then(resolve)
+                .catch((error) => {
+                    if (providerIndex + 1 >= storageProviders.length) {
+                        reject(error);
+                        return;
+                    }
+
+                    runOperation(providerIndex + 1)
+                        .then(resolve)
+                        .catch(reject);
```

Comment:
**gut feeling:** The level of abstraction here obfuscates things. Curious to learn the answers to my last questions though before assuming it is unnecessary.

Reply:
I agree, Marc. From looking at `init()`, it doesn't look like there would ever be two storage providers at once. Either the on-disk one initializes and the in-memory one is removed, or vice versa. It feels a bit over-engineered to assume there would be multiple data providers running at the same time.

Although, if this would allow the in-memory provider to be used for the in-memory cache, then I could understand why `runConsecutively()` would be necessary (but we aren't doing this, so it's still over-engineered).

Reply:
Currently, the `Onyx.init` and `Storage.init` functions are not yet adapted to this, but this was the future plan